pkg/hs/king/lib/Ur/Vere/Pier.hs

```diff
@@ -130,18 +130,36 @@ booted pill lite flags ship boot = do
 -- Resume an existing ship. ----------------------------------------------------
 
 resumed :: (HasPierConfig e, HasLogFunc e)
-        => Serf.Flags
+        => Maybe Word64 -> Serf.Flags
         -> RAcquire e (Serf e, EventLog, SerfState)
-resumed flags = do
+resumed event flags = do
     top    <- view pierPathL
+    tap    <- case event of
+                Nothing -> pure top
+                Just ev -> rio $ do
+                  snap <- getSnapshot top ev
+                  case snap of
+                    Nothing -> pure top
+                    Just sn -> pure sn
+    rio $ logTrace $ displayShow tap
```

Comment:
Minor style nits: generally wouldn't make a big deal out of this, but you asked for style feedback.

The `snap` variable is useless. It doesn't need a name since it's clear what comes out of `getSnapshot`. I would write this with `\case`:

```

    tap    <- case event of
                Nothing -> pure top
                Just ev -> rio $ do
                  getSnapshot top ev >>= \case
                    Nothing -> pure top
                    Just sn -> pure sn
```

The `pure top` bit is repeated. This can be made cleaner with `MaybeT`:

```
import Control.Monad.Trans.Maybe
...
    tap    <- fmap (fromMaybe top) $ rio $ runMaybeT $ do
                  ev <- MaybeT (pure event)
                  MaybeT (getSnapshot top ev)
```

Your code is fine, but it's faster for me to read after these changes:

- Ok, I'm doing something that might fail and I will fall back to `top` in any edge case.
- Ok, first I'll fail if `event` is `Nothing`.
- Ok, Then I'm getting the snapshot, which might also fail.

Reply:
Let me make sure I understand the types:

- `fmap` is in `RAcquire`, so everything after the first `$` needs to be an `RAcquire e (Maybe a)`.
- `rio` takes an `RIO a` and converts to `RAcquire e a`, so we now need an `RIO (Maybe a)`.
- `runMaybeT RIO` takes a `MaybeT RIO a` and converts to `RIO (Maybe a)`, which brings us to the `do` block, which uses to the `MaybeT RIO a` monad.
- We get the event number by creating a `MaybeT RIO a` with `MaybeT` applied to a `RIO (Maybe a)`, so the `pure` is in the `RIO` monad.  This step is necessary since `event` is a `Maybe a`, so this is will short-circuit to `top` here if `event` is `Nothing`.
- Then we get the snapshot by doing the same thing, except instead of `pure` we use `getSnapshot` (because there's actually IO).

Reply:
Yep! That's all correct.

This is concept heavy code, so I might not write this if I was trying to make my code accessible to people less familiar with these idioms. Also, if the logic itself is complicated, sometimes I'll still prefer the explicit-nested-pattern-matching styles.

But, once you're familiar, you mostly don't have to think through this. I'll do basically exactly the process that you wrote down if I made a mistake writing something like this, but when I read it, I already know it type checks so I just filter out most of the conversions in my mind, so I basically just see `fromMaybe top ... rio ... maybeT ... ev <- event ...getSnapshot top ev`.